scripts/keyframes.py

```diff
@@ -0,0 +1,422 @@
+                        help="The command file to read")
+
+    # Add command-line switches for all filters.
+    for filter_class in Filter.__subclasses__():
+        f = filter_class()
+        f.preArgumentParse(parser)
+
+    # Look for a configuration file.  If present, read it to get additional arguments.
+    # Comments and quoting are handled by the shlex module.  The use of a configuration
+    # file allows a user to customize the default behavior of the script as desired.
+    config_args = []
+    # The name of the configuration filename is based on the name of the currently
+    # running script.
+    config_path = getConfigPath(f'.{pathlib.Path(__file__).stem}rc')
+    if config_path is not None:
+        # The shlex module will handle comments and quoting.
+        config_args = shlex.split(config_path.read_text(), comments=True)
+
+    # Parse arguments.
+    args = parser.parse_args(sys.argv[1:] + config_args)
+
+    # Set up logging.
+    logging.basicConfig(stream=sys.stdout, format=LOG_FORMAT, level=args.log_level)
+
+    # If we're using extra arguments, let the user know.
+    if config_args:
+        log.info('using additional arguments from config file %s: %s',
+                 config_path, shlex.join(config_args))
+
+    # Prepare a list of filters that we're actually going to use.
+    filters = []
+    if args.filters is not None:
+        for filter_name in args.filters.split(','):
+            f = getFilterByName(filter_name)
+            filters.append(f)
+            # Let active filters manage their arguments.
+            f.postArgumentParse(args)
+
+    # Read frames from the command file and convert to footprints.
+    # Also collect a dictionary of command counts, which we may write out if desired.
+    command_counts = {}
+    footprints_to_frames = getCommandFileFootprintsToFrames(args.commandfile, command_counts=command_counts)
+
+    # Save the command counts if requested.
+    if args.save_command_counts:
+        with open(args.save_command_counts, mode='wt', encoding='utf-8') as f:
+            log.info('Writing command counts to file %s', args.save_command_counts)
+            json.dump(command_counts, f, sort_keys=True, indent=4)
+
+    # Extract the maximal footprints from the dictionary.
+    maximal_footprints = getMaximalSets(footprints_to_frames.keys())
+
+    # Select a frame from each of our maximal footprints, until we
+    # have the count we want or until we run out of footprints.
+    keyframes = getKeyframes(maximal_footprints, footprints_to_frames, args.n, filters)
+
+    # Give our output.
+    log.info("Keyframes in priority order: %s", ",".join([str(x) for x in keyframes]))
+    keyframes.sort()
+    log.info("Keyframes in sorted order:   %s", ",".join([str(x) for x in keyframes]))
+
+    # Create stub reference images if requested.  The stub reference images essentially
+    # encode the keyframe numbers, which can be used with a normal CI run to generate the
```

Comment:
Seems very coupled to details of internal CI processes.

Reply:
This whole script is tied to internal CI processes - the original point of it was to let an outside vendor create packages that we could "just use" in internal CI.

I'm not at all against making this an internal script - we used such an internal-only script for a long time before a need came for outside vendors to process traces like we do.  If we're going to support outside contributions of test workloads, it'd be nice if we used the same process that we advertise.